core/node/helpers.go

```diff
@@ -0,0 +1,67 @@
+package node
+
+import (
+	"context"
+
+	config "github.com/ipfs/go-ipfs-config"
+	uio "github.com/ipfs/go-unixfs/io"
+	"github.com/jbenet/goprocess"
+	"go.uber.org/fx"
+)
+
+// lifecycleCtx creates a context which will be cancelled when lifecycle stops
+//
+// This is a hack which we need because most of our services use contexts in a
+// wrong way
+func lifecycleCtx(mctx MetricsCtx, lc fx.Lifecycle) context.Context {
+	ctx, cancel := context.WithCancel(mctx)
+	lc.Append(fx.Hook{
+		OnStop: func(_ context.Context) error {
+			cancel()
+			return nil
+		},
+	})
+	return ctx
+}
+
+type lcProcess struct {
+	fx.In
+
+	LC   fx.Lifecycle
+	Proc goprocess.Process
+}
+
+func (lp *lcProcess) Run(f goprocess.ProcessFunc) {
```

Comment:
nit: Maybe just call this `Append`? Not much better but `Run` sounds like something is happening now.